### PR TITLE
cloud-provider: add log options, allow setting logging-format via CLI option

### PR DESCRIPTION
**Pull request by LittleFox94**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

This PR allows users to configure logging format for CCMs based on k8s.io/cloud-provider.

#### Does this PR introduce a user-facing change?

```release-note
* add logging-format option to CCMs based on k8s.io/cloud-provider
```

### CODE REVIEW

**Review by andrewsykim — COMMENTED** [review]
/ok-to-test

**Comment by LittleFox94** [comment]
whoops - sorry for the failed CI run and noise

**Comment by LittleFox94** [comment]
hm.. that failed test looks weird, like not a problem of me?

<details>
<summary>Strange log entry</summary>

```
Terminated (Error - Program process in pidfile '/var/run/docker-ssd.pid', 1 process(es), refused to die. ================================================================================ Done cleaning up after docker in docker. {"component":"entrypoint","error":"wrapped process failed: exit status 2","file":"k8s.io/test-infra/prow/entrypoint/run.go:80","func":"k8s.io/test-infra/prow/entrypoint.Options.Run","level":"error","msg":"Error executing test process","severity":"error","time":"2022-03-24T21:54:43Z"} ) at 2022-03-24 21:54:43 +0000 UTC with exit code 2
--


```
</details>

/retest

**Comment by LittleFox94** [comment]
@andrewsykim can you please take a look at the failing test, if there is something I have to do?

**Comment by LittleFox94** [comment]
Please take another look @andrewsykim, @deads2k or @wlan0 

Tests ran successfully, now only needs another look at the code and hopefully merge :)

Thanks :)

**Comment by LittleFox94** [comment]
/test pull-kubernetes-conformance-kind-ga-only-parallel

looks like a test pod failed, not sure if this is a flake or a "sometimes nodes have problems" thing to live with

**Review by andrewsykim — COMMENTED** [review]
/cc @cici37

**Comment by mrbobbytables** [comment]
/assign @cici37 @cheftako @nckturner 
to get more eyes on it

**Comment by LittleFox94** [comment]
@cici37 we'd like to have logs written in JSON format in our (@Anexia) own CCM and I think all cloud providers would benefit from this option.

JSON logging is available in other components already (https://kubernetes.io/docs/concepts/cluster-administration/system-logs/#json-log-format), this just adds that to CCMs.

**Comment by LittleFox94** [comment]
> Thanks for the PR. Would you mind sharing more on why this is needed? Is this a request from a specific cloud provider? Thank you

do you need any more information from me?

**Comment by LittleFox94** [comment]
I believe this is a subtask of https://github.com/kubernetes/enhancements/issues/1602 ([Structured logging](https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/1602-structured-logging)), allowing CCMs to match the goal "Simplify ingestion of logs into third party logging solutions by adding an option to output logs in the JSON format"

**Review by pohly — APPROVED** [review]
/lgtm

**Comment by LittleFox94** [comment]
/assign @cheftako

has lgtm, only needs last approval now :)

**Comment by LittleFox94** [comment]
This PR is slowly approaching an age of one year, with a multitude of rebases in between and, as I learnt now, rebases will reset the lgtm label..

Would be really great to get this merged, seems like most required review is done already anyway? What's blocking this?

**Comment by pohly** [comment]
Someone who has approval powers for the code has to approve it. It's unfortunate that people ignore assigned PRs, but it happens. In that case, bring it to the SIG which owns the code and ask for approval.

**Review by pohly — APPROVED** [review]
/lgtm

File: staging/src/k8s.io/cloud-provider/app/controllermanager.go
Comment (andrewsykim) [inline staging/src/k8s.io/cloud-provider/app/controllermanager.go:92]: What's the reason for nil feature gate here?

File: staging/src/k8s.io/cloud-provider/app/controllermanager.go
Comment (andrewsykim) [inline staging/src/k8s.io/cloud-provider/app/controllermanager.go:128]: In general, I would like the logging options here to be consistent with kube-controller-manager and this seems to deviate from this: https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-controller-manager/app/controllermanager.go#L155-L162

Comment (andrewsykim) [reply staging/src/k8s.io/cloud-provider/app/controllermanager.go:128]: Ah, actually we are still doing this here: https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-controller-manager/app/options/options.go#L188

**Review by andrewsykim — COMMENTED** [review]
/approve

Thanks and sorry for the delay

**Comment by andrewsykim** [comment]
/hold

(holding for the question on feature gates)

File: staging/src/k8s.io/cloud-provider/app/controllermanager.go
Comment (LittleFox94) [reply staging/src/k8s.io/cloud-provider/app/controllermanager.go:92]: [docs](https://pkg.go.dev/k8s.io/component-base@v0.26.1/logs/api/v1#ValidateAndApply) say "If nil, the default for these features is used", which sounds like a good fit here?

Comment (LittleFox94) [reply staging/src/k8s.io/cloud-provider/app/controllermanager.go:92]: e.g. metrics-server does the same: https://github.com/kubernetes-sigs/metrics-server/blob/master/cmd/metrics-server/app/options/options.go#L59

File: staging/src/k8s.io/cloud-provider/app/controllermanager.go
Comment (LittleFox94) [reply staging/src/k8s.io/cloud-provider/app/controllermanager.go:128]: so this can stay as-is or should I change something?

File: staging/src/k8s.io/cloud-provider/app/controllermanager.go
Comment (andrewsykim) [reply staging/src/k8s.io/cloud-provider/app/controllermanager.go:92]: metrics-server doesn't inherit and default feature gates from Kubernetes though -- I think at the very least we want to use the default feature gate we register here: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/cloud-provider/app/controllermanager.go#L66

Comment (andrewsykim) [reply staging/src/k8s.io/cloud-provider/app/controllermanager.go:92]: `utilfeature.DefaultFeatureGate` is actually better because it doesn't have to be mutable.

File: staging/src/k8s.io/cloud-provider/app/controllermanager.go
Comment (andrewsykim) [reply staging/src/k8s.io/cloud-provider/app/controllermanager.go:128]: I think it can stay as-is

File: staging/src/k8s.io/cloud-provider/app/controllermanager.go
Comment (LittleFox94) [reply staging/src/k8s.io/cloud-provider/app/controllermanager.go]: changed and did a short `go test ./...` below `staging/src/k8s.io/cloud-provider`, but did not yet test any more

Comment (pohly) [reply staging/src/k8s.io/cloud-provider/app/controllermanager.go]: Beware that if a non-nil feature gate instance is passed, then the component-base/logs features must have been added to it.

Comment (LittleFox94) [reply staging/src/k8s.io/cloud-provider/app/controllermanager.go]: @pohly added it to the DefaultMutableFeatureGate, same as `component-base/metrics/features` is added already

Comment (LittleFox94) [reply staging/src/k8s.io/cloud-provider/app/controllermanager.go]: Before I just rebase this PR again, is the way I implemented the feature gate ok?

Comment (pohly) [reply staging/src/k8s.io/cloud-provider/app/controllermanager.go]: Yes, looks good to me.

Comment (LittleFox94) [reply staging/src/k8s.io/cloud-provider/app/controllermanager.go]: thanks :)

so another version prepared, freshly rebased and ready for your eyes @andrewsykim :)

Comment (LittleFox94) [reply staging/src/k8s.io/cloud-provider/app/controllermanager.go]: another ping @andrewsykim

**Review by andrewsykim — COMMENTED** [review]
/approve
/lgtm
/hold cancel

**Pull request merged**
